src/huggingface_hub/hf_api.py

```diff
@@ -9575,6 +9579,75 @@ def list_user_following(self, username: str, token: Union[bool, str, None] = Non
+        the method raises an appropriate exception.
+
+        Args:
+            repo_id (str):
+                The repository to check for access. Format should be `"user/repo_name"`.
+                Example: `"user/my-cool-model"`.
+
+            repo_type (str, optional):
+                The type of the repository. Should be one of `"model"`, `"dataset"`, or `"space"`.
+                If not specified, the default is `"model"`.
+
+            token (Union[bool, str, None], optional):
+                A valid user access token. If not provided, the locally saved token will be used, which is the
+                recommended authentication method. Set to `False` to disable authentication.
+                Refer to: https://huggingface.co/docs/huggingface_hub/quick-start#authentication.
+
+        Raises:
+            RepositoryNotFoundError:
+                Raised if the repository does not exist, is private, or the user does not have access. This can
+                occur if the `repo_id` or `repo_type` is incorrect or if the repository is private but the user
+                is not authenticated.
+
+            GatedRepoError:
+                Raised if the repository exists but is gated and the user is not authorized to access it.
+
+        Example:
+            Check if the user has access to a repository:
+
+            ```python
+            >>> from huggingface_hub import auth_check
+            >>> from huggingface_hub.utils import GatedRepoError, RepositoryNotFoundError
+
+            try:
+                auth_check("user/my-cool-model")
+            except GatedRepoError:
+                # Handle gated repository error
+                print("You do not have permission to access this gated repository.")
+            except RepositoryNotFoundError:
+                # Handle repository not found error
+                print("The repository was not found or you do not have access.")
+            ```
+
+            In this example:
+            - If the user has access, the method completes successfully.
+            - If the repository is gated or does not exist, appropriate exceptions are raised, allowing the user
+            to handle them accordingly.
+        """
+        headers = self._build_hf_headers(token=token)
+        if repo_type is None:
+            repo_type = constants.REPO_TYPE_MODEL
+        if repo_type not in constants.REPO_TYPES:
+            raise ValueError(f"Invalid repo type, must be one of {constants.REPO_TYPES}")
+        path = f"{self.endpoint}/api/{repo_type}s/{repo_id}/auth-check"
+        r = get_session().get(path, headers=headers)
+        try:
+            hf_raise_for_status(r)
+        except GatedRepoError as e:
+            e.append_to_message(" You are not authorized to access this gated repository.")
+            raise
+        except RepositoryNotFoundError as e:
+            e.append_to_message(_AUTH_CHECK_NO_REPO_ERROR_MESSAGE)
+            raise
```

Comment:
```suggestion
        hf_raise_for_status(r)
```

I would not add extra information here. I tried it locally and for a gated repo, you already get:

```
Cannot access gated repo for url https://huggingface.co/api/models/Andrianos/bert-base-greek-punctuation-prediction-finetuned/auth-check.
Access to model Andrianos/bert-base-greek-punctuation-prediction-finetuned is restricted and you are not in the authorized list. Visit https://huggingface.co/Andrianos/bert-base-greek-punctuation-prediction-finetuned to ask for access.
```

and for a missing/private repo you have

```
Repository Not Found for url: https://huggingface.co/api/models/missing/do-not-exist/auth-check.
Please make sure you specified the correct `repo_id` and `repo_type`.
If you are trying to access a private or gated repo, make sure you are authenticated.
```

I don't think either message needs more clarification here.